### PR TITLE
3.3:  Bring two key Agency bug fixes, plus some secondary stuff back to 3.3

### PR DESCRIPTION
**Pull request by matthewvon**
1. includes changes to 14 files using new API for Node.cpp calls that do NOT throw
2. includes CriticalThread reporting for Agency so there is hourly logging if Supervisor dies
3. includes 3rdParty/date because it is in devel and simplifies this back port

### CODE REVIEW

**Comment by jsteemann** [comment]
Adding the date library to a *stable* release here seems overkill to me.
`date` is only used a single time in this entire changeset for generating a log message.
Please just use `std::string TRI_StringTimeStamp(double, bool useLocalTime);` instead and remove the date library again. This will reduce the size of this PR by 99.99%.

**Comment by jsteemann** [comment]
Additionally, I don't like these useless hourly log messages that say "HeartbeatThread ok". 
That provides absolutely no value for end users. Worse, it suggests that if the HeartbeatThread was still there this is exceptional. We should log *failures*, not the normal behavior.

**Comment by jsteemann** [comment]
@matthewvon : can you also adjust the CMakeLists.txt file so that the date library is not put into the include path anymore? Thanks!

**Comment by matthewvon** [comment]
@jsteemann I am retesting.  Please do not consider this ready for merge.  But I am happy to respond to other code comments quickly.

**Comment by matthewvon** [comment]
@jsteemann I am now done.  Please review the hourly reporting of dead threads with Max.  This was added due to a specific customer where logs had been rotated and the clear indication of dead thread was lost.

File: arangod/Cluster/HeartbeatThread.cpp [hunk 1]
```diff
@@ -545,7 +596,7 @@ void HeartbeatThread::runSingleServer() {
           // preemtily remove the transient entry from the agency
           _agency.setTransient(transientPath, VPackSlice::emptyObjectSlice(), 0);
```
Comment (graetzer) [inline arangod/Cluster/HeartbeatThread.cpp]: this seems to be incomplete. In devel `_agency.setTransient(transientPath, VPackSlice::emptyObjectSlice(), 0);` is not there anymore

**Comment by matthewvon** [comment]
@graetzer and @jsteemann :  I backed out every change to HeartbeatThread.cpp except the dead thread reporting.  Removing the hourly "HearbeatThread ok" message is a simple, single line comment if required (this corresponds to Monty Python's line "Not dead yet").

My intent is to create an independent branch for the other changes to HeartbeatThread.cpp that are most useful for 3.3 based upon my prior PRs.  This method will exclude the code that caused Simon's concern.

If code passes your review I will merge current 3.3 HEAD and get a Jenkins test scheduled.

**Comment by kvahed** [comment]
LGTM

**Comment by matthewvon** [comment]
@graetzer and @jsteemann I am going to learn how to start a Jenkins build Friday morning.  I think that will complete this branch unless you have other comments.

**Comment by matthewvon** [comment]
@jsteemann 

PR has passed Jenkins build.  Is it ready for merge?

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/527/

**Comment by jsteemann** [comment]
I have discussed with @neunhoef and we were both happy to change the log level for the "HeartbeatThread ok" message from INFO to DEBUG.
I also added a missing mutex for the deadThreads handling and the deadThreadsPosted variable.

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/535/

**Comment by jsteemann** [comment]
Fine for me, but I asked Max or Kaveh for review before merging back.

**Comment by kvahed** [comment]
Looks fine

**Pull request merged**